sonarr/indexer.go

```diff
@@ -0,0 +1,196 @@
+package sonarr
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"sort"
+	"strconv"
+
+	"golift.io/starr"
+)
+
+type Indexer struct {
+	EnableAutomaticSearch   bool           `json:"enableAutomaticSearch"`
+	EnableInteractiveSearch bool           `json:"enableInteractiveSearch"`
+	EnableRss               bool           `json:"enableRss"`
+	SupportsRss             bool           `json:"supportsRss"`
+	SupportsSearch          bool           `json:"supportsSearch"`
+	Priority                int64          `json:"priority"`
+	ID                      int64          `json:"id,omitempty"`
+	ConfigContract          string         `json:"configContract"`
+	Implementation          string         `json:"implementation"`
+	ImplementationName      string         `json:"implementationName"`
+	InfoLink                string         `json:"infoLink"`
+	Name                    string         `json:"name"`
+	Protocol                string         `json:"protocol"`
+	Tags                    []*starr.Value `json:"tags"`
+	Fields                  []*starr.Field `json:"fields"`
+}
+
+const bpIndexer = APIver + "/indexer"
+
+var (
+	IndexerFieldsString   = []string{"additionalParameters", "apiKey", "apiPath", "baseUrl", "captchaToken", "cookie", "passkey", "username"}
+	IndexerFieldsBool     = []string{"allowZeroSize", "rankedOnly"}
+	IndexerFieldsInt      = []string{"delay", "minimumSeeders", "seedCriteria.seasonPackSeedTime", "seedCriteria.seedTime"}
+	IndexerFieldsFloat    = []string{"seedCriteria.seedRatio"}
+	IndexerFieldsIntSlice = []string{"animeCategories", "categories"}
+)
```

Comment:
Can't these be struct members?

Reply:
Don't know. They are the value that `name` parameter should have to map the `value` parameter, I checked them on the source code since I couldn't find docs.

My idea was actually to implement them using a function since we can't have static slices. Something like:
```go
func getIndexerFBool() []string {
	return []string{"allowZeroSize", "rankedOnly"}
}
```

Reply:
Thank you for the extra context and json payloads. That helps. 

Questions:
- Is the input data for fields always just name and value? 
- The output data is always _more_ than name and value?

I'm going to think about this for a bit.

_The following is probably not useful yet (I'm still thinking about this, and the below paragraph is just my thoughts that may be accurate)._

What comes to mind first is that we should probably disconnect the input and output struct from the input data. Meaning, we use a different struct to hit sonar than what we expect our library consumers to put into this library. We return a slightly different format for The output. We could even add methods to update these known field types, and perhaps a method that takes in an `interface{}` to catch any we miss or new ones that get added in the future. We could accomplish that with hidden fields that get processed when the payload is passed in, similar to what you're doing now.

Example:
```go
indexer := &Index{Name: "indexer1"}
indexer.SetFieldString("cookie", "string")
indexer.SetFieldBool("allowZeroSize", true)
indexer.SetFieldCategories([]int{})
indexer.SetFieldAnimeCategories([]int{})
```

Reply:
I sort of feel like if we know what all these field names (and types) are we should represent them in a `struct`. The struct can be 'processed' into the interface{} data before doing the http request. Reverse that for the output: put the interface{} data back into a struct.